---
# Feel free to add content and custom Front Matter to this file.
# To modify the layout, see https://jekyllrb.com/docs/themes/#overriding-theme-defaults

layout: home
title: Home
---

<!-- ## Home -->


## RECUR: Identifying recurrent amino acid substitutions from multiple sequence alignments
<br>

![RECUR method workflow](./assets/images/RECUR_workflow_figure.png)

<div align="center" style="font-size: 18px; font-weight: bold;">
  Figure 1: The RECUR workflow
</div>
<br>



The required input is either a protein or codon multiple sequence alignment (in FASTA format) and a defined outgroup species or clade. The output of RECUR is a list of recurrent amino acid substitutions, that have occurred in the inferred phylogeny (file suffix: `.recur.tsv`). Outputs of intermediate steps, i.e. model selection, tree inference, ancestral state reconstruction and site substitution matrices, can be found in the .recur output directory.


## How to cite?


To cite RECUR, please cite both the following:

[1] recur ref
 
[2] *Minh BQ, Schmidt HA, Chernomor O, Schrempf D, Woodhams MD, von Haeseler A, Lanfear R.* 2020. **IQ-TREE 2: New Models and Efficient Methods for Phylogenetic Inference in the Genomic Era.** Molecular Biology and Evolution 37:1530-1534. [![DOI:10.1093/molbev/msaa015](https://raw.githubusercontent.com/OrthoFinder/RECUR/main/docs/images/doi-badge.svg)](https://doi.org/10.1093/molbev/msaa015)


<p style="font-size: 18px;">
  If you want to cite an exemplar of RECURâ€™s usage, please cite:
</p>


[3] *Robbins EHJ, Kelly S.* 2024. **Widespread adaptive evolution in angiosperm photosystems provides insight into the evolution of photosystem II repair.** The Plant Cell 37. [![ DOI: 10.1093/plcell/koae281](https://raw.githubusercontent.com/OrthoFinder/RECUR/main/docs/images/doi-badge.svg)](https://doi.org/10.1093/plcell/koae281)

